Rewrite of Windows x64 value type ABI to support DirectX wrapper libraries on JIT and full AOT.

lateralusX opened the pull request:
Current implementation used different value type ABI’s depending on if call
was a managed->native, native->managed or managed->managed call. The implementation
was also overcomplicated since it was a tweaked copy from System V x64 value type
ABI with scenarios and conditions that wouldn’t apply to the Windows x64 value
type ABI. Having a uniform value type ABI when possible regardless of managed->native,
native->managed or managed->managed calls simplifies scenarios and reduce complexity.

While all the above are legitimate reasons the primary driving factor behind
the rewrite was a different issue. There are .NET wrapper libraries around DirectX API
used by game engines and other libraries that uses a pattern using a call into a stub
managed method that gets patched directly in the MSIL to a calli that will end up in a
native DirectX vtable method. In current ABI we can’t detect scenario like this, so arguments
are passed as if the call was a managed->managed call causing issues for value types. Since DirectX
uses value types a lot (vectors, matrixes etc.), this was a blocker getting these technologies
running on Windows using DirectX at all. Another approach could be to change the behavior of the
library, but this would require a major redesign of the library and since several of these libraries
are mature, that wasn’t a realistic option, especially since MS .NET runtime handles the libraries as is
without modifications.

The rewrite makes sure Windows x64 value type ABI follow the platform value type ABI in all scenarios.

This PR also include additional fixes needed in order to get full AOT working for DirectX  in full AOT mode on Windows x64.

All regressions tests pass on both JIT and full AOT on Windows x64.

alexrp commented on mono/mini/mini-amd64.c:
```diff
@@ -428,206 +428,217 @@ collect_field_info_nested (MonoClass *klass, StructFieldInfo *fields, int index,
-	if (!sig->pinvoke) {
-		int n = mono_class_value_size (klass, NULL);
+inline gboolean
+allocate_return_register_for_valyetype_win64 (ArgInfo *arg_info, ArgumentClass arg_class, guint32 arg_size, guint32 *current_int_reg, guint32 *current_float_reg)
```
typo

alexrp commented on mono/mini/mini-amd64.c:
```diff
@@ -1225,6 +1236,10 @@ mono_arch_cpu_optimizations (guint32 *exclude_mask)
 	}
 
+#ifdef TARGET_WIN32
+	*exclude_mask |= MONO_OPT_SIMD;
```
This change is unexplained.

lateralusX replied:
The current SIMD doesn't support the argument used by a LD_ADDR to be of type OP_VTARG_ADDR. This will now be used for value types > 8 or of size 3,5,6,7 as dictated by windows x64 value type ABI. Since OP_VTARG_ADDR needs to be resolved in mono_spill_global_vars and the SIMD implementation optimize away the LD_ADDR in load_simd_vreg, that will cause an error in mono_spill_global_vars since incorrect opcode will now have a reference to an argument that won't be fully resolved. I looked into just resolving the optimization but it turns out that there are several constructs that could end up with different opcodes having the same problem emitted by the SIMD optimizations. I discussed this with @vargaz and the recommendation was to disable this on win x64 for now and look into this later, if needed.

vargaz replied:
Just add a small comment about it.

alexrp commented on mono/cil/cil-opcodes.xml:
```diff
@@ -319,4 +319,5 @@
 <opcode name="mono_calli_extra_arg" input="VarPop" output="VarPush" args="InlineSig" o1="0xF0" o2="0x18" flow="call" />
 <opcode name="mono_lddomain" input="Pop0" output="PushI" args="InlineNone" o1="0xF0" o2="0x19" flow="next" />
 <opcode name="mono_atomic_store_i4" input="PopI+PopI" output="Push0" args="InlineI" o1="0xF0" o2="0x1A" flow="next" />
+<opcode name="mono_get_last_error" input="Pop0" output="PushI" args="InlineI" o1="0xF0" o2="0x1B" flow="next" />
```
The opcode takes no args, so should be `InlineNone`.

kumpera replied:
Can't use use the existing TLS loading opcode/IR op here?

lateralusX replied:
Win32 last error is not in the TLS but in the TEB (thread environment block) at a different offset 0x068, compared to the TLS data slots that are located on offset 0x1480 or 0x1780 depending on TLS key.

vargaz commented on mono/mini/mini-amd64.c:
```diff
@@ -428,206 +428,217 @@ collect_field_info_nested (MonoClass *klass, StructFieldInfo *fields, int index,
+
+			/* Only RAX/XMM0 should be used to return valyetype. */
+			assert ((arg_info->pair_regs[0] == AMD64_RAX && arg_info->pair_regs[1] == ArgNone) || (arg_info->pair_regs[0] == AMD64_XMM0 && arg_info->pair_regs[1] == ArgNone));
+
```
Extra line

vargaz commented on mono/mini/mini-amd64.c:
```diff
@@ -428,206 +428,217 @@ collect_field_info_nested (MonoClass *klass, StructFieldInfo *fields, int index,
+			assert ((arg_info->pair_regs[0] == AMD64_RAX && arg_info->pair_regs[1] == ArgNone) || (arg_info->pair_regs[0] == AMD64_XMM0 && arg_info->pair_regs[1] == ArgNone));
+
+		} else {
+
```
Ditto.

vargaz commented on mono/mini/mini-amd64.c:
```diff
@@ -428,206 +428,217 @@ collect_field_info_nested (MonoClass *klass, StructFieldInfo *fields, int index,
+	assert (klass != NULL && arg_info != NULL && type != NULL && arg_class != NULL && arg_size != NULL);
+	
+	if (pinvoke == TRUE) {
+		/* Calculate argument class type and size of marshalled type. */
```
if (pinvoke)

vargaz commented on mono/mini/mini-amd64.c:
```diff
@@ -428,206 +428,217 @@ collect_field_info_nested (MonoClass *klass, StructFieldInfo *fields, int index,
+		arg_info->storage = ArgValuetypeInReg;
+		arg_info->pair_storage [0] = arg_info->pair_storage [1] = ArgNone;
+	}
+	else
```
This (and other places) should use K&R indentation ie. } else {

vargaz commented on mono/mini/mini-amd64-gsharedvt.c:
```diff
@@ -148,11 +149,13 @@ get_arg_slots (ArgInfo *ainfo, int **out_slots, gboolean is_source_argument)
 		break;
 	case ArgValuetypeInReg:
+	case ArgValuetypeAddrInIReg:
 		nsrc = ainfo->nregs;
```
This doesn't look correct.

lateralusX replied:
In ArgValuetypeAddrInIReg ainfo->nregs will always be 1 so I reused the case to reduce code duplication. I can split into a separate case with nsrc set to 1 if that makes the intent clearer, but with some code duplication.

vargaz commented on mono/mini/mini-amd64-gsharedvt.c:
```diff
@@ -148,11 +149,13 @@ get_arg_slots (ArgInfo *ainfo, int **out_slots, gboolean is_source_argument)
 			src [i] = map_reg (ainfo->pair_regs [i]);
 		break;
+	case ArgValuetypeAddrOnStack:
 	case ArgOnStack:
```
Ditto.

lateralusX replied:
In ArgValuetypeAddrOnStack ainfo->arg_size will always be 8 so I reused the case to reduce code duplication. I can split into a separate case with nsrc set to 1 if that makes the intent clearer, but with some code duplication.

vargaz commented on mono/metadata/marshal.c:
```diff
@@ -422,7 +422,8 @@ mono_delegate_to_ftnptr (MonoDelegate *delegate)
 	return delegate->delegate_trampoline;
```
Looks ok.

vargaz commented on mono/metadata/marshal.c:
```diff
@@ -7456,19 +7456,27 @@ mono_marshal_emit_native_wrapper (MonoImage *image, MonoMethodBuilder *mb, MonoM
+			mono_mb_emit_native_call (mb, get_last_error_sig, GetLastError);
+			mono_mb_emit_icall (mb, mono_marshal_set_last_error_windows);
+		}
+		else
```
Use K&R indentation.

lateralusX commented:
Fixed smaller formatting issues based on feedback and split case statement in mini-amd64-gsharedvt.c to make the intent clearer. Thanks for the feedback!

vargaz commented:
LGTM. Would be nice to squash the changes.
